File: ios/Classes/StreamSessionManager.swift
import Flutter
import MWDATCamera
import MWDATCore

class StreamSessionManager {
    private var streamSession: StreamSession?
    private var renderer: VideoFrameRenderer?
    private var stateListenerToken: AnyListenerToken?
    private var videoFrameListenerToken: AnyListenerToken?
    private var errorListenerToken: AnyListenerToken?
    private var photoDataListenerToken: AnyListenerToken?
    private var sessionId: String = ""

    private let streamStateHandler: StreamStateStreamHandler
    private let videoFrameInfoHandler: VideoFrameInfoStreamHandler
    private let streamErrorsHandler: StreamErrorsStreamHandler
    private let textureRegistry: FlutterTextureRegistry

    private var pendingPhotoCompletion: ((Result<[String: Any], Error>) -> Void)?

    init(
        textureRegistry: FlutterTextureRegistry,
        streamStateHandler: StreamStateStreamHandler,
        videoFrameInfoHandler: VideoFrameInfoStreamHandler,
        streamErrorsHandler: StreamErrorsStreamHandler
    ) {
        self.textureRegistry = textureRegistry
        self.streamStateHandler = streamStateHandler
        self.videoFrameInfoHandler = videoFrameInfoHandler
        self.streamErrorsHandler = streamErrorsHandler
    }

    func startSession(resolution: StreamingResolution, frameRate: Int, wearables: WearablesInterface) -> [String: Any] {
        stopSession()

        let videoRenderer = VideoFrameRenderer(textureRegistry: textureRegistry)
        renderer = videoRenderer

        let config = StreamSessionConfig(
            videoCodec: .raw,
            resolution: resolution,
            frameRate: frameRate
        )
        let deviceSelector = AutoDeviceSelector(wearables: wearables)
        let session = StreamSession(streamSessionConfig: config, deviceSelector: deviceSelector)
        streamSession = session
        sessionId = "\(Int(Date().timeIntervalSince1970 * 1000))"

        stateListenerToken = session.statePublisher.listen { [weak self] state in
            Task { @MainActor [weak self] in
                self?.streamStateHandler.eventSink?(
                    Converters.streamSessionStateToString(state)
                )
            }
        }

        videoFrameListenerToken = session.videoFramePublisher.listen { [weak self] videoFrame in
            Task { @MainActor [weak self] in
                guard let self else { return }
                self.renderer?.renderFrame(videoFrame)
                if let image = videoFrame.makeUIImage() {
                    self.videoFrameInfoHandler.eventSink?([
                        "width": Int(image.size.width * image.scale),
                        "height": Int(image.size.height * image.scale),
                        "timestampMs": Int(Date().timeIntervalSince1970 * 1000)
                    ])
                }
            }
        }

        errorListenerToken = session.errorPublisher.listen { [weak self] error in
            Task { @MainActor [weak self] in
                self?.streamErrorsHandler.eventSink?([
                    "code": Converters.streamErrorToString(error),
                    "message": "\(error)"
                ])
            }
        }

        photoDataListenerToken = session.photoDataPublisher.listen { [weak self] photoData in
            Task { @MainActor [weak self] in
                guard let self else { return }
                let data = photoData.data
                if let uiImage = UIImage(data: data) {
                    let result: [String: Any] = [
                        "bytes": FlutterStandardTypedData(bytes: data),
                        "width": Int(uiImage.size.width * uiImage.scale),
                        "height": Int(uiImage.size.height * uiImage.scale)
                    ]
                    self.pendingPhotoCompletion?(.success(result))
                    self.pendingPhotoCompletion = nil
                }
            }
        }

        Task {
            await session.start()
        }

        return [
            "sessionId": sessionId,
            "textureId": videoRenderer.textureId
        ]
    }

    func stopSession() {
        stateListenerToken = nil
        videoFrameListenerToken = nil
        errorListenerToken = nil
        photoDataListenerToken = nil

        if let session = streamSession {
            Task {
                await session.stop()
            }
        }
        streamSession = nil
        renderer?.dispose()
        renderer = nil
        pendingPhotoCompletion = nil
    }

    func capturePhoto(completion: @escaping (Result<[String: Any], Error>) -> Void) {
        guard let session = streamSession else {
            completion(.failure(NSError(domain: "MetaWearables", code: -1, userInfo: [NSLocalizedDescriptionKey: "No active session"])))
            return
        }
        pendingPhotoCompletion = completion
        session.capturePhoto(format: .jpeg)
    }

    func getSessionId() -> String { sessionId }
}

File: android/src/main/kotlin/com/meta/wearable/flutter/StreamSessionManager.kt
package com.meta.wearable.flutter

import android.app.Application
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.meta.wearable.dat.camera.StreamSession
import com.meta.wearable.dat.camera.startStreamSession
import com.meta.wearable.dat.camera.types.PhotoData
import com.meta.wearable.dat.camera.types.StreamConfiguration
import com.meta.wearable.dat.camera.types.StreamSessionState
import com.meta.wearable.dat.camera.types.VideoQuality
import com.meta.wearable.dat.core.Wearables
import com.meta.wearable.dat.core.selectors.AutoDeviceSelector
import io.flutter.view.TextureRegistry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.io.ByteArrayOutputStream

class StreamSessionManager(
    private val application: Application,
    private val textureRegistry: TextureRegistry,
    private val scope: CoroutineScope,
    private val streamStateHandler: WearablesStreamHandlers.StreamStateHandler,
    private val videoFrameInfoHandler: WearablesStreamHandlers.VideoFrameInfoHandler,
    private val streamErrorsHandler: WearablesStreamHandlers.StreamErrorsHandler
) {
    private var streamSession: StreamSession? = null
    private var renderer: VideoFrameRenderer? = null
    private var videoJob: Job? = null
    private var stateJob: Job? = null
    private var sessionId: String = ""

    fun startSession(
        videoQuality: VideoQuality,
        frameRate: Int
    ): Map<String, Any> {
        stopSession()

        val textureEntry = textureRegistry.createSurfaceTexture()
        val videoRenderer = VideoFrameRenderer(textureEntry)
        renderer = videoRenderer

        val deviceSelector = AutoDeviceSelector()
        val config = StreamConfiguration(videoQuality = videoQuality, frameRate)
        val session = Wearables.startStreamSession(application, deviceSelector, config)
        streamSession = session
        sessionId = System.currentTimeMillis().toString()

        videoJob = scope.launch(Dispatchers.Main) {
            session.videoStream.collect { videoFrame ->
                videoRenderer.renderFrame(videoFrame)
                videoFrameInfoHandler.eventSink?.success(
                    mapOf(
                        "width" to videoFrame.width,
                        "height" to videoFrame.height,
                        "timestampMs" to System.currentTimeMillis()
                    )
                )
            }
        }

        stateJob = scope.launch(Dispatchers.Main) {
            session.state.collect { state ->
                streamStateHandler.eventSink?.success(
                    Converters.streamSessionStateToString(state)
                )
            }
        }

        return mapOf(
            "sessionId" to sessionId,
            "textureId" to videoRenderer.textureId
        )
    }

    fun stopSession() {
        videoJob?.cancel()
        videoJob = null
        stateJob?.cancel()
        stateJob = null
        streamSession?.close()
        streamSession = null
        renderer?.release()
        renderer = null
    }

    suspend fun capturePhoto(): Map<String, Any>? {
        val session = streamSession ?: return null
        val result = session.capturePhoto()
        var photoResult: Map<String, Any>? = null

        result.onSuccess { photoData ->
            val bitmap = when (photoData) {
                is PhotoData.Bitmap -> photoData.bitmap
                is PhotoData.HEIC -> {
                    val byteArray = ByteArray(photoData.data.remaining())
                    photoData.data.get(byteArray)
                    BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
                }
            }
            if (bitmap != null) {
                val stream = ByteArrayOutputStream()
                bitmap.compress(Bitmap.CompressFormat.JPEG, 90, stream)
                val bytes = stream.toByteArray()
                stream.close()
                photoResult = mapOf(
                    "bytes" to bytes,
                    "width" to bitmap.width,
                    "height" to bitmap.height
                )
            }
        }

        return photoResult
    }

    fun getSessionId(): String = sessionId
}
